torch/csrc/jit/passes/batch_mm.cpp

```diff
@@ -0,0 +1,173 @@
+#include "torch/csrc/jit/passes/batch_mm.h"
+
+#include "torch/csrc/jit/passes/dead_code_elimination.h"
+#include "torch/csrc/jit/interned_strings.h"
+#include "torch/csrc/utils/functional.h"
+
+#include <ATen/ATen.h>
+#include <algorithm>
+#include <unordered_map>
+
+namespace torch { namespace jit {
+
+// This pass looks for trees in the graph, where leaves are mm ops, and the inner
+// vertices are add nodes. Once we have such a tree they can be reduced to two
+// concats and a single mm (basically into a single multiply of a wide matrix, with
+// a tall matrix).
+// Such patterns show up mostly in backward of RNNs, since the derivative of many
+// uses of matrix multiplies with same weights forms exactly such a tree
+// (note that it's usually also highly imbalanced i.e. has O(n) depth).
```

Comment:
Could this optimization be bad for memory usage if you're doing higher order gradients?

Reply:
Possible, but if we add a "register allocation" pass (for no-op concats), then they will only improve things